source/opt/loop_descriptor.cpp

```diff
@@ -103,6 +104,179 @@ bool Loop::IsBasicBlockInLoopSlow(const BasicBlock* bb) {
+  cfg->RemoveNonExistingEdges(loop_header_->id());
+  // Update the loop descriptors.
+  if (HasParent()) {
+    GetParent()->AddBasicBlock(loop_preheader_);
+    context->GetLoopDescriptor(fn)->SetBasicBlockToLoop(loop_preheader_->id(),
+                                                        GetParent());
+  }
+
+  context->InvalidateAnalysesExceptFor(
+      builder.GetPreservedAnalysis() |
+      ir::IRContext::Analysis::kAnalysisLoopAnalysis |
+      ir::IRContext::kAnalysisCFG);
+
+  return loop_preheader_;
+}
+
+void Loop::SetLatchBlock(BasicBlock* latch) {
+#ifndef NDEBUG
+  assert(latch->GetParent() && "The basic block does not belong to a function");
+
+  latch->ForEachSuccessorLabel([this](uint32_t id) {
+    assert((!IsInsideLoop(id) || id == GetHeaderBlock()->id()) &&
+           "A predecessor of the continue block does not belong to the loop");
+  });
+#endif  // NDEBUG
+  assert(IsInsideLoop(latch) && "The continue block is not in the loop");
+
+  SetLatchBlockImpl(latch);
+  if (IsStructured()) {
+    UpdateLoopMergeInst();
+  }
+}
+
+void Loop::SetMergeBlock(BasicBlock* merge) {
+#ifndef NDEBUG
+  assert(merge->GetParent() && "The basic block does not belong to a function");
+  CFG& cfg = *merge->GetParent()->GetParent()->context()->cfg();
+
+  for (uint32_t pred : cfg.preds(merge->id())) {
+    assert(IsInsideLoop(pred) &&
+           "A predecessor of the merge block does not belong to the loop");
+  }
+#endif  // NDEBUG
+  assert(!IsInsideLoop(merge) && "The merge block is in the loop");
+
+  SetMergeBlockImpl(merge);
+  if (IsStructured()) {
+    UpdateLoopMergeInst();
+  }
+}
+
+void Loop::GetExitBlocks(IRContext* context,
+                         std::unordered_set<uint32_t>* exit_blocks) const {
+  ir::CFG* cfg = context->cfg();
+
+  for (uint32_t bb_id : GetBlocks()) {
+    const spvtools::ir::BasicBlock* bb = cfg->block(bb_id);
+    bb->ForEachSuccessorLabel([exit_blocks, this](uint32_t succ) {
+      if (!IsInsideLoop(succ)) {
+        exit_blocks->insert(succ);
+      }
+    });
+  }
+}
+
+bool Loop::IsLCSSA() const {
+  IRContext* context = GetHeaderBlock()->GetParent()->GetParent()->context();
+  ir::CFG* cfg = context->cfg();
+  opt::analysis::DefUseManager* def_use_mgr = context->get_def_use_mgr();
+
+  std::unordered_set<uint32_t> exit_blocks;
+  GetExitBlocks(context, &exit_blocks);
+
+  for (uint32_t bb_id : GetBlocks()) {
+    for (Instruction& insn : *cfg->block(bb_id)) {
+      // All uses must be either:
+      //  - In the loop;
+      //  - In an exit block and in a phi instruction.
+      if (!def_use_mgr->WhileEachUser(
+              &insn, [&exit_blocks, context, this](ir::Instruction* use) {
+                BasicBlock* parent = context->get_instr_block(use);
+                assert(parent && "Invalid analysis");
+                if (IsInsideLoop(parent)) return true;
+                if (use->opcode() != SpvOpPhi) return false;
+                return !!exit_blocks.count(parent->id());
```

Comment:
Remove the !!.

Reply:
Yes the lambda must return a bool, my old C reflex is to do this ... (that's a common idiom to do 0/1 normalization which also do a conversion to bool type in C++).

Another approach is to specify the type in the lambda declaration: `(ir::Instruction* use) -> bool`.